src/style.js

```diff
@@ -356,15 +352,27 @@ const getFrameClip = memoizeClip((clipPath, context, dimensions) => {
     .attr("height", height - marginTop - marginBottom);
 });
 
-const getProjectionClip = memoizeClip((clipPath, context) => {
-  const {projection} = context;
-  if (!projection) throw new Error(`the "sphere" clip option requires a projection`);
-  clipPath.append("path").attr("d", geoPath(projection)({type: "Sphere"}));
-});
+function memoizeGeo(clip) {
+  const geoClips = new WeakMap();
+  return (geo, scales) => {
+    if (!geoClips.has(geo)) geoClips.set(geo, clip(geo, scales));
+    return geoClips.get(geo);
+  };
+}
+
+const getGeoClip = memoizeGeo((geo, scales) =>
+  memoizeClip((clipPath, context) => {
+    const {projection} = context;
+    if (!projection && geo.type === "Sphere") throw new Error(`clipping to the sphere requires a projection`);
+    clipPath.append("path").attr("d", geoPath(projection ?? xyProjection(scales))(geo));
```

Comment:
That’s a lot of plumbing of `scales` so we can call `xyProjection` here. I wonder if instead we should have a (bound) method or getter on the `context` so that we only need to provide the `context` here. Or maybe we _always_ have a `context.projection` and it’s initialized from the _x_ and _y_ scales if needed (but I think there are other places where we test if there’s a projection, so I doubt that will work).

Reply:
maybe be a context.geopath getter?

Reply:
context.path it is